jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/IOResources.java

```diff
@@ -0,0 +1,525 @@
+        Path path = resource.getPath();
+        if (path != null)
+        {
+            RangedPathContentSource contentSource = new RangedPathContentSource(path, bufferPool, first, length);
+            if (bufferSize > 0)
+            {
+                contentSource.setBufferSize(bufferSize);
+                contentSource.setUseDirectByteBuffers(direct);
+            }
+            return contentSource;
+        }
+
+        // Try an optimization for MemoryResource.
+        if (resource instanceof MemoryResource memoryResource)
+            return new ByteBufferContentSource(ByteBuffer.wrap(memoryResource.getBytes()));
+
+        // Fallback to InputStream.
+        try
+        {
+            InputStream inputStream = resource.newInputStream();
+            if (inputStream == null)
+                throw new IllegalArgumentException("Resource does not support InputStream: " + resource);
+            RangedInputStreamContentSource contentSource = new RangedInputStreamContentSource(inputStream, bufferPool, first, length);
+            if (bufferSize > 0)
+            {
+                contentSource.setBufferSize(bufferSize);
+                contentSource.setUseDirectByteBuffers(direct);
+            }
+            return contentSource;
+        }
+        catch (IOException e)
+        {
+            throw new RuntimeIOException(e);
+        }
+    }
+
+    /**
+     * <p>Gets an {@link InputStream} with the contents of a resource.</p>
+     * <p>The resource must not be a directory, must exists and there must return non-null to {@link Resource#newInputStream()}.</p>
+     *
+     * @param resource the resource from which to get an {@link InputStream}.
+     * @return the {@link InputStream}.
+     * @throws IllegalArgumentException if the resource is a directory or does not exist or {@link Resource#newInputStream()} returns null.
+     */
+    public static InputStream asInputStream(Resource resource) throws IllegalArgumentException
+    {
+        if (resource.isDirectory() || !resource.exists())
+            throw new IllegalArgumentException("Resource must exist and cannot be a directory: " + resource);
+        try
+        {
+            InputStream inputStream = resource.newInputStream();
+            if (inputStream == null)
+                throw new IllegalArgumentException("Resource does not support InputStream: " + resource);
+            return inputStream;
+        }
+        catch (IOException e)
+        {
+            throw new RuntimeIOException(e);
+        }
+    }
+
+    /**
+     * <p>Performs an asynchronous copy of the contents of a resource to a sink, using the given buffer pool and buffer characteristics.</p>
+     * <p>The resource must not be a directory, must exists and there must be
+     * a way to access its contents.</p>
+     * <p>Multiple optimized methods are used to access the resource's contents but if they all fail,
+     * {@link Resource#newInputStream()} is used as a fallback.</p>
```

Comment:
Isn't `ContentSource` the fallback? I don't see any call to `Resource.newInputStream`?

Reply:
`Resource.newInputStream` is the fallback of the fallback, but you're right: `asContentSource()` is what should be documented as the fallback.